### PR TITLE
Supports UTF-8 file with BOM.

### PR DESCRIPTION
**Pull request by tifish**
The code may be a bit rough:
- Does it break any convension?
- Does it cover all cases?
- The icon may be inappropriate. I just pick one from Android Studio.

### CODE REVIEW

**Comment by gsantner** [comment]
As of icon, that is OK - thats just what everybody else did so far here :D

File: app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java [hunk 1]
```diff
@@ -171,6 +171,11 @@ private AlertDialog.Builder makeDialog(final File basedir, final boolean allowCr
                     final byte[] templateContents = getTemplateContent(templateSpinner, basedir, f.getName(), encryptCheckbox.isChecked());
                     shareUtil.writeFile(f, false, (arg_ok, arg_fos) -> {
                         try {
+                            if (appSettings.isEditorAddBomToNewFile()) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java:174]: Maybe just make it a toggle on the newfile dialog, not a general settings option. so when you create a new file you tick that on or off if you want to.

what do you think?

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java [hunk 2]
```diff
@@ -60,6 +61,37 @@ public static String readTextFileFast(final File file) {
         return "";
     }
 
+    public static Pair<String, Boolean> readTextFileFastWithBOM(final File file) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/util/FileUtils.java:64]: preferably we don't stack read methods onTop onTop onTop with different features.   
We already had that and the current focus on simplifying and debloating.

So what I rather suggest is, have one such function, and have a Pair<String, Map> as result. And in that Map any metadata can be put. With the first case here being BOM information.

What do you think.

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java [hunk 3]
```diff
@@ -177,8 +209,13 @@ public static byte[] readCloseBinaryStream(final InputStream stream) {
         return baos.toByteArray();
     }
 
-    public static boolean writeFile(final File file, final byte[] data) {
+    public static boolean writeFile(final File file, final byte[] data, boolean withBOM) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/util/FileUtils.java:212]: when you agree with the metadata comment, you could pass it in here again and so withBom information gets auto extracted from that `Map`.

**Review by gsantner — CHANGES_REQUESTED** [review]
^

File: app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java (hunk 1 above)
Comment (tifish) [reply app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java:174]: That's a good idea.

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java (hunk 2 above)
Comment (tifish) [reply app/src/main/java/net/gsantner/opoc/util/FileUtils.java:64]: I'm not familiar with Java. Thanks for your advice.

**Comment by tifish** [comment]
I committed all changes. See if I missed something.

**Comment by gsantner** [comment]
OK should I try it on the weekend?

File: app/src/main/java/net/gsantner/markor/model/Document.java [hunk 4]
```diff
@@ -257,14 +262,21 @@ public synchronized boolean saveContent(final Context context, final String cont
...
                         fos.write(contentAsBytes);
                         fos.flush();
                     } catch (Exception ignored) {
                     }
                 });
                 success = true;
             } else {
-                success = FileUtils.writeFile(_file, contentAsBytes);
+                final Map<String, Object> options = new HashMap<>(1);
+                options.put("withBom", _withBom);
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/model/Document.java:278]: plz make the "withBom" key a `final static` constant

**Comment by tifish** [comment]
Yeah, tell me if anything I can do.

File: app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java [hunk 5]
```diff
@@ -171,6 +177,11 @@ private AlertDialog.Builder makeDialog(final File basedir, final boolean allowCr
...
                     shareUtil.writeFile(f, false, (arg_ok, arg_fos) -> {
                         try {
+                            if (appSettings.getNewFileDialogLastUsedUtf8Bom()) {
+                                arg_fos.write(0xEF);
+                                arg_fos.write(0xBB);
+                                arg_fos.write(0xBF);
+                            }
                             if (templateContents != null && (!f.exists() || f.length() < ShareUtil.MIN_OVERWRITE_LENGTH)) {
                                 arg_fos.write(templateContents);
                             }
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/NewFileDialog.java:187]: We should get wrid of own file writing implementation here and instead use FileUtils, what is used at editor as well.

But that is a topic of it's own.

**Review by gsantner — APPROVED** [review]

**Comment by harshad1** [comment]
@tifish FYI, The bom test in Document #189 is causing a crash for me on start. The bug does not appear in debug mode, just release mode.

Also, why are we returning a map from readTextFileFast insead of a simple boolean?

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java [hunk 6]
```diff
@@ -44,20 +47,41 @@ public class FileUtils {
     // Used on methods like copyFile(src, dst)
     private static final int BUFFER_SIZE = 4096;
 
-    public static String readTextFileFast(final File file) {
+    public final static String WITH_BOM = "withBom";
+
+    public static Pair<String, Map<String, Object>> readTextFileFast(final File file) {
+        Map<String, Object> info = new HashMap<>(1);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/util/FileUtils.java:53]: This, why does it need to be a hashmap? From my reading, a boolean will do, right?

File: app/src/main/java/net/gsantner/opoc/util/FileUtils.java (hunk 2 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/util/FileUtils.java]: Ah, I see why we chose the map approach :)

**Pull request merged**